micro-rdk/src/esp32/camera.rs

```diff
@@ -1,117 +1,265 @@
+                fb_count,
+                grab_mode,
+                fb_location,
+                sccb_i2c_port,
             },
-            last_grab: t.now(),
-        }
-    }
-    pub fn setup(&self) -> Result<(), CameraError> {
-        let ret = (unsafe { crate::esp32::esp_idf_svc::sys::esp_camera_init(&self.config) })
-            as crate::esp32::esp_idf_svc::sys::esp_err_t;
-        let ret = crate::esp32::esp_idf_svc::sys::EspError::convert(ret);
-        ret.map_err(|e| CameraError::CameraInitError(e.into()))
-    }
-    pub fn get_cam_frame(&self) -> Option<*mut crate::esp32::esp_idf_svc::sys::camera_fb_t> {
-        let ptr = (unsafe { crate::esp32::esp_idf_svc::sys::esp_camera_fb_get() })
-            as *mut crate::esp32::esp_idf_svc::sys::camera_fb_t;
-        if ptr.is_null() {
-            None
-        } else {
-            Some(ptr)
-        }
-    }
-    pub fn return_cam_frame(
-        &self,
-        frame: Option<*mut crate::esp32::esp_idf_svc::sys::camera_fb_t>,
-    ) {
-        if let Some(ptr) = frame {
-            unsafe { crate::esp32::esp_idf_svc::sys::esp_camera_fb_return(ptr) }
-        }
+            last_grab: Instant::now(),
+        };
+
+        let ret =
+            (unsafe { esp_camera_init(&cam.config) }) as crate::esp32::esp_idf_svc::sys::esp_err_t;
+        EspError::convert(ret).map_err(|e| CameraError::InitError(e.into()))?;
+        Ok(Arc::new(Mutex::new(cam)))
     }
-    pub fn debug_print_fb(&self, frame: &Option<*mut crate::esp32::esp_idf_svc::sys::camera_fb_t>) {
-        if let Some(ptr) = frame {
-            let ptr = ptr as &*mut crate::esp32::esp_idf_svc::sys::camera_fb_t;
-            unsafe {
-                println!();
-                info!("camera buf size {}", (*(*ptr)).len);
-                println!();
-                for i in 0..(*(*ptr)).len {
-                    print!("{:02X}", (*(*(*ptr)).buf.offset(i as isize)));
-                    if i > 0 && i % 80 == 0 {
-                        println!();
-                    }
-                }
-            }
-        }
+
+    fn get_frame(&mut self) -> Result<Esp32CameraFrameBuffer, CameraError> {
+        self.last_grab = Instant::now();
+        let frame = Esp32CameraFrameBuffer::get().ok_or_else(|| CameraError::FailedToGetImage)?;
+        Ok(frame)
     }
 }
+
 impl Camera for Esp32Camera {
-    fn get_frame(&mut self, mut buffer: BytesMut) -> Result<BytesMut, CameraError> {
-        if let Some(ptr) = self.get_cam_frame() {
-            let buf = unsafe {
-                let buf = (*ptr).buf;
-                let len = (*ptr).len as usize;
-                core::slice::from_raw_parts(buf, len)
-            };
-            if buf.len() > buffer.capacity() {
-                self.return_cam_frame(Some(ptr));
-                return Err(CameraError::CameraFrameTooBig);
+    fn get_image(&mut self, mut buffer: BytesMut) -> Result<BytesMut, CameraError> {
+        if let Ok(frame) = self.get_frame() {
```

Comment:
either remove the Ok(frame) or change get_frame to only return Option<> otherwise you are masking  the error returned by get_frame